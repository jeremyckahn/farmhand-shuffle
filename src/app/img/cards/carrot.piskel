{"modelVersion":2,"piskel":{"name":"carrot","description":"","fps":12,"height":24,"width":24,"layers":["{\"name\":\"Layer 1\",\"opacity\":1,\"frameCount\":1,\"chunks\":[{\"layout\":[[0]],\"base64PNG\":\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAA+klEQVRIS82UOw7CMBBEnQrRUHIBDpiKU1BxQC5ASYNSBdmJzXo/3jGSJVJa9jzPeDZTGPxNg/XDnwDuxzU5nd/dF8IPRIgGsNb37NsAepi7AF35DjgkugDFowkMkPPPMC0WA+oDojgV5o+dhY0S4AA6MDQmp10eYKvn5kLOJFDbFmB93Q5F9HRdahB3wR0CNa0Acb+AAMNnOSjij+cSLmfmhMfViMoFxEuqkFxxZyY0gHp7Ctmj+p5tQJoA7fbkLbwGpmfmm8zsq0dG/gBGi0RzfhXOraUOkriSNfbPkmMoIiq3L33viMLQl28ADKClpa5DTehSZJuHAz7EuYYZ/S2Q+AAAAABJRU5ErkJggg==\"}]}"]}}