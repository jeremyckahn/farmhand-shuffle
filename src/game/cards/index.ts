import { v4 as uuid } from 'uuid'

import { ICard, Instance } from '../types'

export * from './crops'
export * from './events'
export * from './water'

export const instantiate = <T = ICard>(card: T): T & Instance => {
  return { ...card, instanceId: uuid() }
}
