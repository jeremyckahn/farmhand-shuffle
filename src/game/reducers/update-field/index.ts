import { IField, IGame, IPlayer } from '../../types'
import { updatePlayer } from '../update-player'

export const updateField = (
  game: IGame,
  playerId: IPlayer['id'],
  newFieldProperties: Partial<IField>
) => {
  const { field } = game.table.players[playerId]

  game = updatePlayer(game, playerId, {
    field: { ...field, ...newFieldProperties },
  })

  return game
}
