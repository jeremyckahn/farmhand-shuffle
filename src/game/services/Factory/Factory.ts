import { v4 as uuid } from 'uuid'

import { initialPlayerFunds } from '../../config'
import { IField, IGame, IPlayer, ITable } from '../../types'

export class Factory {
  static buildField(overrides: Partial<IField> = {}): IField {
    return {
      crops: [],
      ...overrides,
    }
  }

  static buildPlayer(overrides: Partial<IPlayer> = {}): IPlayer {
    return {
      id: uuid(),
      funds: initialPlayerFunds,
      deck: [],
      hand: [],
      discardPile: [],
      field: Factory.buildField(overrides?.field),
      ...overrides,
    }
  }

  static buildTable(overrides: Partial<ITable> = {}): ITable {
    return {
      players: {},
      communityFund: 100,
      ...overrides,
    }
  }

  static buildGame(overrides: Partial<IGame> = {}): IGame {
    return {
      table: Factory.buildTable(overrides?.table),
      ...overrides,
    }
  }
}
