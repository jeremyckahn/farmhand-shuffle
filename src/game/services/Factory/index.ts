import { v4 as uuid } from 'uuid'

import { INITIAL_PLAYER_FUNDS } from '../../config'
import { ICrop, IField, IGame, IPlayedCrop, IPlayer, ITable } from '../../types'

export class Factory {
  static buildField(overrides: Partial<IField> = {}): IField {
    return {
      crops: [],
      ...overrides,
    }
  }

  static buildPlayer(overrides: Partial<IPlayer> = {}): IPlayer {
    return {
      id: uuid(),
      funds: INITIAL_PLAYER_FUNDS,
      deck: [],
      hand: [],
      discardPile: [],
      field: Factory.buildField(overrides?.field),
      ...overrides,
    }
  }

  static buildTable(overrides: Partial<ITable> = {}): ITable {
    return {
      players: {},
      communityFund: 100,
      ...overrides,
    }
  }

  static buildGame(overrides: Partial<IGame> = {}): IGame {
    const table = Factory.buildTable(overrides?.table)
    const [currentPlayerId = null] = Object.keys(table.players)

    return {
      table,
      currentPlayerId,
      ...overrides,
    }
  }

  static buildPlayedCrop({ id }: ICrop): IPlayedCrop {
    return {
      id,
      waterCards: 0,
    }
  }
}
