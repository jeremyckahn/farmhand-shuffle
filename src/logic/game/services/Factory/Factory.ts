import { v4 as uuid } from 'uuid'

import { initialPlayerFunds } from '../../config'
import { Field, Game, Player, Table } from '../../types/index'

export class Factory {
  static buildField(): Field {
    return {
      crops: [],
    }
  }

  static buildPlayer(id = uuid()): Player {
    return {
      id,
      funds: initialPlayerFunds,
      deck: [],
      hand: [],
      discardPile: [],
      field: Factory.buildField(),
    }
  }

  static buildTable(): Table {
    const player1 = Factory.buildPlayer()

    return {
      // TODO: This is a WIP implementation. Design for supporting for multiple
      // players is not yet determined.
      players: { [player1.id]: player1 },

      communityFund: 0,
    }
  }

  static buildGame(): Game {
    return {
      table: Factory.buildTable(),
    }
  }
}
