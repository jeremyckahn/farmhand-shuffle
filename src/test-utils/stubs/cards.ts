import { carrot, instantiate, pumpkin, rain, water } from '../../game/cards'
import { CropInstance, EventInstance, WaterInstance } from '../../game/types'

export const stubCarrot: CropInstance = instantiate(carrot)
export const stubPumpkin: CropInstance = instantiate(pumpkin)
export const stubWater: WaterInstance = instantiate(water)
export const stubRain: EventInstance = instantiate(rain)
