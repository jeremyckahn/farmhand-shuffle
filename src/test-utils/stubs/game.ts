import { IGame } from '../../game/types'

import { stubTable } from './table'

export const stubGame = (): IGame => {
  return {
    table: stubTable(),
  }
}
