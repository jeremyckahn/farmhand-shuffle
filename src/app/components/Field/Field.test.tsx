import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Factory } from '../../../game/services/Factory'
import { updateField } from '../../../game/reducers/update-field'
import { stubGame } from '../../../test-utils/stubs/game'
import { carrot, pumpkin } from '../../../game/cards'

import {
  Field,
  FieldProps,
  rotationTransform,
  selectedCardLabel,
  unselectedCardLabel,
} from '.'

let gameStub = stubGame()
const opponentPlayerId = Object.keys(gameStub.table.players)[1]

const fieldCrop1 = carrot
const fieldCrop2 = pumpkin

const cropsStub = [
  { ...Factory.buildPlayedCrop(fieldCrop1), waterCards: 1 },
  { ...Factory.buildPlayedCrop(fieldCrop2), waterCards: 3 },
]

gameStub = updateField(gameStub, gameStub.sessionOwnerPlayerId, {
  crops: cropsStub,
})

gameStub = updateField(gameStub, opponentPlayerId, {
  crops: cropsStub,
})

const StubField = (overrides: Partial<FieldProps>) => {
  return (
    <Field
      game={gameStub}
      playerId={gameStub.sessionOwnerPlayerId}
      {...overrides}
    />
  )
}

describe('Field', () => {
  test('renders crop cards for self', () => {
    render(<StubField />)

    const playedCrops = screen.getAllByLabelText(unselectedCardLabel)

    for (const playedCrop of playedCrops) {
      const { transform } = getComputedStyle(playedCrop)
      expect(transform).toMatchSnapshot()
    }
  })

  test('renders crop cards for opponent upside-down', async () => {
    render(<StubField playerId={opponentPlayerId} />)

    const playedCrops = screen.getAllByLabelText(unselectedCardLabel)

    for (const playedCrop of playedCrops) {
      const { transform } = getComputedStyle(playedCrop)
      expect(transform).toMatchSnapshot()
      expect(transform).toContain(rotationTransform)
    }
  })

  test('clicking a card selects it', async () => {
    render(<StubField />)

    const [playedCrop1, ...restPlayedCrops] =
      screen.getAllByLabelText(unselectedCardLabel)

    await userEvent.click(playedCrop1)

    const { transform: playedCrop1Transform } = getComputedStyle(playedCrop1)

    expect(playedCrop1Transform).toMatchInlineSnapshot(
      `"translateX(512px) translateY(359px) scale(1.25)"`
    )
    expect(playedCrop1).toHaveAttribute('aria-label', selectedCardLabel)

    for (const playedCrop of restPlayedCrops) {
      const { transform } = getComputedStyle(playedCrop)
      expect(transform).toMatchSnapshot()
      expect(playedCrop).toHaveAttribute('aria-label', unselectedCardLabel)
    }
  })

  test("clicking an opponent's card selects it", async () => {
    render(<StubField playerId={opponentPlayerId} />)

    const [playedCrop1] = screen.getAllByLabelText(unselectedCardLabel)

    await userEvent.click(playedCrop1)

    const { transform } = getComputedStyle(playedCrop1)

    expect(transform).toMatchInlineSnapshot(
      `"translateX(512px) translateY(409px) scale(1.25)"`
    )
    expect(transform).not.toContain(rotationTransform)
  })

  test('losing focus resets the card selection', async () => {
    render(<StubField />)

    const [playedCrop1] = screen.getAllByLabelText(unselectedCardLabel)

    await userEvent.click(playedCrop1)

    await waitFor(() => {
      ;(document.activeElement as HTMLElement).blur()
    })

    const { transform } = getComputedStyle(playedCrop1)
    expect(transform).toMatchInlineSnapshot(`""`)
    expect(playedCrop1).toHaveAttribute('aria-label', unselectedCardLabel)
  })

  test('supports tab navigation', async () => {
    render(<StubField />)

    const [playedCrop1, playedCrop2] =
      screen.getAllByLabelText(unselectedCardLabel)
    await userEvent.click(playedCrop1)

    await waitFor(() => {
      userEvent.keyboard('{Tab}')
    })

    const { transform: card1Transform } = getComputedStyle(playedCrop1)
    const { transform: card2Transform } = getComputedStyle(playedCrop2)

    expect(card1Transform).toMatchInlineSnapshot(`""`)
    expect(card2Transform).toMatchInlineSnapshot(
      `"translateX(512px) translateY(359px) scale(1.25)"`
    )
    expect(playedCrop1).toHaveAttribute('aria-label', unselectedCardLabel)
    expect(playedCrop2).toHaveAttribute('aria-label', selectedCardLabel)
  })

  test('focus can be escaped', async () => {
    render(<StubField />)

    const [playedCrop1] = screen.getAllByLabelText(unselectedCardLabel)

    await userEvent.click(playedCrop1)

    await waitFor(() => {
      userEvent.keyboard('{Escape}')
    })

    const { transform: card1Transform } = getComputedStyle(playedCrop1)
    expect(card1Transform).toMatchInlineSnapshot(`""`)
    expect(document.activeElement).toBe(document.body)
  })
})
