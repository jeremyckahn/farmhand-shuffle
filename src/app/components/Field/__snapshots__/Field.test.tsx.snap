// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field clicking a card selects it 2`] = `""`;

exports[`Field renders crop cards for opponent upside-down 1`] = `"rotate(180deg)"`;

exports[`Field renders crop cards for opponent upside-down 2`] = `"rotate(180deg)"`;

exports[`Field renders crop cards for self 1`] = `""`;

exports[`Field renders crop cards for self 2`] = `""`;
